### ISSUE TITLE
Setup Angular2 & TypeScript in a ASP.​NET Core project using Visual Studio

### ISSUE THREAD

**Issue by JuergenGutsch** [JuergenGutsch/blog#56]
Written on 08.08.2016 08:21:27 

URL: http://asp.net-hacker.rocks/2016/08/08/setup-angular2-typescript-aspnetcore-in-visualstudio.html

**Comment by JuergenGutsch** [comment]
Comment written by **Vahid N.** on **08.08.2016 09:38:07**

<p>One day you will need to add routing and server side services and at that time, <br><a href="https://www.nuget.org/packages/Microsoft.AspNetCore.SpaServices/" rel="nofollow noopener" title="https://www.nuget.org/packages/Microsoft.AspNetCore.SpaServices/">https://www.nuget.org/packa...</a><br>can help to provide a fallback <br>routes.MapSpaFallbackRoute("spa-fallback", new { controller = "Home", action = "Index" });</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **08.08.2016 09:53:27**

<p>Hi Vahid, <br>Many thanks for that hint. Will have a deeper look into that package :)<br>At the first glimpse, this just seems to be needed, if you use all the MVC stuff too. I don't us it, if I create a SPA.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Hector Miguel Asiain Palomo** on **08.08.2016 21:46:05**

<p>Hello, regarding this: " [..] but VS shows a yellow exclemation mark because of any arror. Until yet, I didn't figure out what is going wrong here." The package generating the conflict is lite-server, or better said a dependency of it. I unistall it and the yellow mark is gone, I won't be using it anyway at this point.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **09.08.2016 08:00:05**

<p>Hello Hector, many thanks for that hint. I'll try to remove it too and update that post.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Hector Miguel Asiain Palomo** on **10.08.2016 19:06:51**

<p>We are not alone on this, more info about the issue:<br><a href="https://github.com/aspnet/Tooling/issues/479" rel="nofollow noopener" title="https://github.com/aspnet/Tooling/issues/479">https://github.com/aspnet/T...</a><br>BTW it is a really good post, very useful, thanks.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **10.08.2016 20:53:18**

<p>Thanks. Will follow the discussion on that issue :)<br>Glad you like that post.<br>BTW I need to approve comments with links in it. That's why it wasn't shown.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **11.08.2016 00:04:07**

<p>This is easily one of the most clearly-written, clean, and straight-to-the-point examples of adding basline Angular2 into a VS2015 .NET Core project. I especially like your no-nonsense gulp script (there are some truly bad ones out there just to move all those files to wwwroot).</p><p>One thing you might want to clarify -- there should be an "app" folder added to the wwwroot folder (I think?), and this is where the main.ts and app.component.ts files go. Took me a few build attempts to sort that out. In my project, at least, there was no app folder initially.</p><p>I'm actually trying to retrofit a basic admin interface into an existing Core API project. After following these instructions, the Angular part isn't working yet. I get a series of TS2304 errors -- Cannot find name Promise, Map, MapConstructor, SetConstructor, and Set. However, es6-promise does show up in the npm Dependencies folder (and I'm running Chrome which shouldn't need it).</p><p></p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **11.08.2016 00:35:21**

<p>Well that didn't take long. The build is fixed thanks to <a href="http://stackoverflow.com/a/36323136/152997" rel="nofollow noopener" title="http://stackoverflow.com/a/36323136/152997">http://stackoverflow.com/a/...</a> ... The tsconfig.json should reference ES6, not ES5. (35 years of programming and we're back to spaghetti code. SMH.)</p><p>One more recommendation -- specify where to put systemjs.config.js. As written (based on index.html) it goes in wwwroot. Personally I think it belongs in app or maybe a scripts folder though.</p><p>There are also typos in systemsjs.config.js -- it specifies "lib" whereas your gulp copies the files to "libs" (plural).</p><p>On top of that I'm getting a lot of 404s trying to read various files from the libs folder.</p><p>At first @angular stuff wasn't coming through. I changed "@angular":"libs/@angular" in the systemjs.config.js since I noticed gulp didn't reproduce the top @angular folder and it cleared those up, but then I started getting 404s on rxjs.</p><p>I don't really know gulp yet but it looks like the correct folder hierarchy isn't being reproduced. I'd definitely like to see the fix for this because, as I noted, your gulp file is pretty clean compared to the hack jobs I see posted in most places elsewhere.</p><p>Edit: Thought we needed to add { base: 'node_modules' } after the src list, but that just causes other errors when things like "reflect.js" are referenced from libs instead of their origin folder in node_modules. I assume there must be a standard way of dealing with this but it's beyond me right now...</p><p>Overall still pretty happy with this though.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **11.08.2016 01:04:11**

<p>The fix for the other 404s is related to the corrected folder structure (well, "complete" folder structure, I don't know if it's correct as per The Angular Way)... the polyfill section in index.html needs to change to:</p><p>    &lt;script src="libs/core-js/client/shim.min.js"&gt;&lt;/script&gt;<br>    &lt;script src="libs/zone.js/dist/zone.js"&gt;&lt;/script&gt;<br>    &lt;script src="libs/reflect-metadata/reflect.js"&gt;&lt;/script&gt;<br>    &lt;script src="libs/systemjs/dist/system.src.js"&gt;&lt;/script&gt;</p><p>And with that, it runs under VS2015 Update 3 with .NET Core 1.0 release and Angular2 RC5.</p><p>(Edit -- note this references an earlier comment that is awaiting moderation, I guess probably because I included a stackoverflow link.)</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **11.08.2016 07:39:26**

<p>Hi, <br>many thanks for all of your comments. :) <br>I'll update this post in the next days to to be more clear in some points and to fix some possible errors. I think I also need to update this post to Angular RC5. <br>Unfortunately they changed a lot while the product is RC (same mistake as Microsoft did with ASP.NET Core from RC1 to RC2) and they will change a lot more until release<br><a href="http://angularjs.blogspot.ch/2016/08/angular-2-rc5-ngmodules-lazy-loading.html" rel="nofollow noopener" title="http://angularjs.blogspot.ch/2016/08/angular-2-rc5-ngmodules-lazy-loading.html">http://angularjs.blogspot.c...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Rodrigo Basniak** on **13.08.2016 15:06:37**

<p>Hi, I tried to follow your steps but when running the npm install I get the following output on the output window:</p><p>PATH=.\node_modules\.bin;C:\Program Files (x86)\Microsoft Visual Studio 14.0\Web\External;%PATH%;C:\Program Files (x86)\Microsoft Visual Studio 14.0\Web\External\git<br>"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Web\External\npm.CMD" install<br>npm WARN install Couldn't install optional dependency: Unsupported<br>&gt; dotnetpro-ecollector@1.0.0 postinstall C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4<br>&gt; typings install &amp;&amp; gulp restore<br>module.js:327<br>    throw err;<br>    ^<br>Error: Cannot find module 'popsicle'<br>    at Function.Module._resolveFilename (module.js:325:15)<br>    at Function.Module._load (module.js:276:25)<br>    at Module.require (module.js:353:17)<br>    at require (internal/module.js:12:17)<br>    at Object.&lt;anonymous&gt; (C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\popsicle-retry\dist\index.js:4:18)<br>    at Module._compile (module.js:397:26)<br>    at Object.Module._extensions..js (module.js:404:10)<br>    at Module.load (module.js:343:32)<br>    at Function.Module._load (module.js:300:12)<br>    at Module.require (module.js:353:17)<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\gaze\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\glob-watcher\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\glogg\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\gulplog\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\has-gulplog\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\hawk\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\http-errors\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\loud-rejection\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\serve-index\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\<a href="http://socket.io" rel="nofollow noopener" title="socket.io">socket.io</a>-parser\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\sparkles\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\spdx-correct\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\spdx-exceptions\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\spdx-expression-parse\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\spdx-license-ids\package.json'<br>npm WARN ENOENT ENOENT: no such file or directory, open 'C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\node_modules\statuses\package.json'<br>npm WARN EPACKAGEJSON dotnetpro-ecollector@1.0.0 No description<br>npm WARN EPACKAGEJSON dotnetpro-ecollector@1.0.0 No repository field.<br>npm ERR! Windows_NT 10.0.10586<br>npm ERR! argv "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Web\\External\\node.exe" "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\Web\\External\\node_modules\\npm\\bin\\npm-cli.js" "install"<br>npm ERR! node v5.4.1<br>npm ERR! npm  v3.3.4<br>npm ERR! code ELIFECYCLE<br>npm ERR! dotnetpro-ecollector@1.0.0 postinstall: `typings install &amp;&amp; gulp restore`<br>npm ERR! Exit status 1<br>npm ERR! <br>npm ERR! Failed at the dotnetpro-ecollector@1.0.0 postinstall script 'typings install &amp;&amp; gulp restore'.<br>npm ERR! This is most likely a problem with the dotnetpro-ecollector package,<br>npm ERR! not with npm itself.<br>npm ERR! Tell the author that this fails on your system:<br>npm ERR!     typings install &amp;&amp; gulp restore<br>npm ERR! You can get their info via:<br>npm ERR!     npm owner ls dotnetpro-ecollector<br>npm ERR! There is likely additional logging output above.<br>npm ERR! Please include the following file with any support request:<br>npm ERR!     C:\Users\rbasniak\Documents\Visual Studio 2015\Projects\WebApplication4\src\WebApplication4\npm-debug.log</p><p>Do you have any idea of what could be wrong?</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Murat Mercan** on **15.08.2016 04:57:37**

<p>am I the only one who uses Visual studio to debug typescript, uses breakpoints and step into/over the lines, please let me know if I am doing wrong but without<br>  meta: { '*.js': { scriptLoad: true } } <br>in systemjs.config.js config I am not be able to debug my typescript in Visual studio</p>

**Comment by JuergenGutsch** [comment]
Comment written by **John Born** on **15.08.2016 23:15:19**

<p>First thing I've checked after hitting this page was the RC version used. I will have a good read if you actually update it to RC5. Many thanks!</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Igor Soloydenko** on **17.08.2016 01:11:11**

<p>If you have fixed it. Could you please provide a sample of a minimal working solution in GitHub?</p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **17.08.2016 01:19:11**

<p>Fortunately it looks like RC5 should be the final changes.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **17.08.2016 01:22:17**

<p>My Core project is already too complicated to easily unwind, but I also set up a bare-bones .NET 4.5.2 VS2015 project and all the important files should be the same. I'll set it up tomorrow and post a link here.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jens Cappelle** on **18.08.2016 16:54:56**

<p>Hi, This might be a dumb question but i'm pretty new to both the <a href="http://asp.net" rel="nofollow noopener" title="asp.net">asp.net</a> scene (been working in it a month and a half now) and Angular(started today). But I can't figure out the part about 'typings.config' do i create a file with that name? and where should i add it? to my src folder, to my wwwroot folder, to my typings folder.. Thanks for this awesome guide btw! saw a lot of guides but didn't like most of them and others were targeting a totally different objective. Cudos</p>

**Comment by JuergenGutsch** [comment]
Comment written by **TruthSayer** on **18.08.2016 18:47:01**

<p>So close yet so far. <br>Can't believe no one on the planet has a  recent working .net core/angular2 site running.<br>10 years making .net websites and have now spent 3 days now trying to  get a "Hello world" displayed on a single page.<br>Thought this was the demo that would show me how build one but there are too many simple errors that beginners in things like npm, gulp, angular etc just can't get over/figure out, having never seen a running error free .net core 'starter' site with those technologies  used. <br>As far as i can see you have maybe 14 days to find and complete a demo like this before everything, npm, gulp, angular2 etc is rewritten to make the demo code obsolete and impossible to  get running.</p><p>Does anyone know or have a sample hello world site done in recent weeks with .net core, npm, gulp, &amp; angular2?</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 20:50:31**

<p>Hi Murat, I also never got this working in Visual Studio. To debug TypeScript, I use the Chrome development tools.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 20:53:26**

<p>Hi, <br>so it doesn't work for you? <br>Currently I'm in vacation, but I'll create a sample, if I'm back at home.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 20:55:28**

<p>I will, if I'm back at home. I'm in vacation now</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 20:57:47**

<p>Yes, will do it in about a week, if I'm back at home</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 21:17:30**

<p>Hi Jens, the typings.config should be located in the root of the project. glad you like that post :)</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **19.08.2016 21:27:12**

<p>The most common issue is, that you don't use the console in administration mode. Open a console in administration mode. <br>Another solution could be to install typings and gulp globally on your machine<br>&gt; npm install typings gulp -g</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Rodrigo Basniak** on **19.08.2016 23:11:14**

<p>That would be awesome if you could. Your tip on running npm as administrator worked for the restore, but I finished following your steps and my application is not working too. I get a eternal loading page. So a working sample would be super duper helpful, because it's exactly like TS said, not a single tutorial on the internet is working right now, the ones that are not outdated give the forever loading screen. Maybe people who have experience on them may fight through the details and make it work, but for someone totally new on web, it's being a real pain to make asp core and angular work together in Visual Studio.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Murat Mercan** on **20.08.2016 15:24:32**

<p>Sorry I am actually debugging with Visual studio from beta version (angular and <a href="http://asp.net" rel="nofollow noopener" title="asp.net">asp.net</a> core beta)<br>this is the <br>tsconfig.json<br>{<br>  "compilerOptions": {<br>    "noImplicitAny": false,<br>    "noEmitOnError": true,<br>    "removeComments": false,<br>    "sourceMap": true,<br>    "target": "es5",<br>    "experimentalDecorators": true,<br>    "emitDecoratorMetadata": true,<br>    "module": "system",<br>    "moduleResolution": "node",<br>    "outDir": "../wwwroot/app"</p><p>  },<br>  "exclude": [<br>    "node_modules",<br>    "wwwroot"<br>  ]<br>}</p><p>and my systemjs.config.js file<br>  meta: { '*.js': { scriptLoad: true } } is important</p><p>/**<br> * System configuration for Angular 2 samples<br> * Adjust as necessary for your application needs.<br> */<br>(function (global) {<br>    // map tells the System loader where to look for things<br>    var map = {<br>        'app': 'app', // 'dist',<br>        '@angular': 'lib/angular2',<br>        'angular2-in-memory-web-api': 'lib/angular2-in-memory-web-api',<br>    };<br>    // packages tells the System loader how to load when no filename and/or no extension<br>    var packages = {<br>        'app': { main: 'main.js', defaultExtension: 'js' },<br>        'angular2-in-memory-web-api': { main: 'index.js', defaultExtension: 'js' },<br>    };<br>    var ngPackageNames = [<br>      'common',<br>      'compiler',<br>      'core',<br>      'forms',<br>      'http',<br>      'platform-browser',<br>      'platform-browser-dynamic',<br>      'router',<br>      'router-deprecated',<br>      'upgrade',<br>    ];<br>    // Individual files (~300 requests):<br>    function packIndex(pkgName) {<br>        packages['@angular/' + pkgName] = { main: 'index.js', defaultExtension: 'js' };<br>    }<br>    // Bundled (~40 requests):<br>    function packUmd(pkgName) {<br>        packages['@angular/' + pkgName] = { main: pkgName + '.umd.js', defaultExtension: 'js' };<br>    }<br>    // Most environments should use UMD; some (Karma) need the individual index files<br>    var setPackageConfig = System.packageWithIndex ? packIndex : packUmd;<br>    // Add package entries for angular packages<br>    ngPackageNames.forEach(setPackageConfig);<br>    var config = {<br>        map: map,<br>        packages: packages,<br>        meta: { '*.js': { scriptLoad: true } }<br>    };<br>    System.config(config);<br>})(this);</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Igor Soloydenko** on **21.08.2016 05:02:48**

<p>Thanks a ton sir!</p>

**Comment by JuergenGutsch** [comment]
Comment written by **MV10** on **22.08.2016 14:02:58**

<p>I'm not sure how my typings.json file got named typings.config, but with VS / Typescript integration you don't really need typings.json or a stand alone typings folder at the project-level, as far as I can tell. I think that's the only issue I've found with that setup as it exists in git right now.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Igor Soloydenko** on **23.08.2016 08:34:16**

<p>Angular2 worked for me only after I rewrote this section as following:</p><p>```<br>    var map = {<br>        'app': 'app',<br>        // 'rxjs': 'libs/rxjs',<br>        // '@angular': 'libs/@angular'<br>        '@angular': 'libs',<br>        'rxjs': 'libs'<br>    };<br>```</p>

**Comment by JuergenGutsch** [comment]
Comment written by **TruthSayer** on **23.08.2016 13:54:39**

<p>It seems to be a mix of Visual Studio  &amp; Typing. <br>This error seems common Cannot find module @angular/core'; on line<br>import {Component} from '@angular/core';</p><p>This fix of pasting this into the  .xproj file doesn't work even on clean starter projects.</p><p>&lt;propertygroup condition="'$(Configuration)' == 'Debug'"&gt;  <br>    &lt;typescripttarget&gt;ES5&lt;/typescripttarget&gt;  <br>    &lt;typescriptjsxemit&gt;None&lt;/typescriptjsxemit&gt;  <br>    &lt;typescriptcompileonsaveenabled&gt;True&lt;/typescriptcompileonsaveenabled&gt;  <br>    &lt;typescriptnoimplicitany&gt;False&lt;/typescriptnoimplicitany&gt;  <br>    &lt;typescriptmodulekind&gt;System&lt;/typescriptmodulekind&gt;  <br>    &lt;typescriptremovecomments&gt;False&lt;/typescriptremovecomments&gt;  <br>    &lt;typescriptoutfile/&gt;  <br>    &lt;typescriptoutdir/&gt;  <br>    &lt;typescriptgeneratesdeclarations&gt;False&lt;/typescriptgeneratesdeclarations&gt;  <br>    &lt;typescriptnoemitonerror&gt;True&lt;/typescriptnoemitonerror&gt;  <br>    &lt;typescriptsourcemap&gt;True&lt;/typescriptsourcemap&gt;  <br>    &lt;typescriptmaproot/&gt;  <br>    &lt;typescriptsourceroot/&gt;  <br>    &lt;typescriptexperimentaldecorators&gt;True&lt;/typescriptexperimentaldecorators&gt;  <br>    &lt;typescriptemitdecoratormetadata&gt;True&lt;/typescriptemitdecoratormetadata&gt;  <br>&lt;/propertygroup&gt;</p>

**Comment by JuergenGutsch** [comment]
Comment written by **nbl** on **08.09.2016 16:11:58**

<p>I don't see yet the benefit of doing all this in Visual Studio. There is no integration with controller action methods. Instead of Index.html if we could engineer all of this in a View that is returned by the controller then we are truly integrating the two technologies.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **gviuytgcv** on **21.09.2016 00:26:37**

<p>Hi There, thanks for putting this together! I'm a bit confused as to where and how the actual Angular 2 js or ts files or whatever are referenced in the index.html? I see</p><p>"&lt;-- 1. Load libraries --&gt; "</p><p>in the index.html but there are no references to the actual Angular files? I'm trying to setup a project and the Angular 2 npm modules has a PILE of folders and files.. not sure what is supposed to be referenced. thanks.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **gviuytgcv** on **21.09.2016 00:29:40**

<p>feeling your pain. Started with .net 1.0.. I must really suck because I'm at least 5 days in. :)</p>

**Comment by JuergenGutsch** [comment]
Comment written by **TruthSayer** on **21.09.2016 00:44:45**

<p>That was a month ago. I'm now on non working demo number 7. Nothing works error free. <br>Typescript is a pile of hacked together pile of junk that gets rewritten every goddamn week. Looking now for someone who will take $10 amazon vouchers to answer simple questions about why these demos don't work.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **21.09.2016 07:06:43**

<p>It is a pain to use visual studio in this case, specially the TypeSCript configuration. Tepending on your size of your project, I would propose separate the .NET &amp; Visual Studio part from the Angular2 part into two different projects using two different IDEs or try VS Code: <a href="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html" rel="nofollow noopener" title="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html">http://asp.net-hacker.rocks...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **21.09.2016 07:09:37**

<p>Hi Rich, <br>The Angular files are loaded automatically with system.js. This is configured in the systemjs.config.js</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **21.09.2016 07:13:41**

<p>Hello, <br>the benefit is just doing all in VS (some peaople have to do it like this) and to host the Angular2 App inside an ASP.NET web. That is all, but as I read here in the comments, VS could be a pain in some cases. this is why I wrote the same thing using dotnet CLI and VS Code: <a href="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html" rel="nofollow noopener" title="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html">http://asp.net-hacker.rocks...</a><br>Aniway, ASP.NET is reduced to just use the Web API to provide data and it didn't use any classic ASP.NET MVC  stuff. You're also able to separate the client part and the server part into two different projects, which is the best solution for larger projects.<br>Cheers,</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **21.09.2016 07:26:19**

<p>Thanks :)</p>

**Comment by JuergenGutsch** [comment]
Comment written by **gviuytgcv** on **21.09.2016 18:22:11**

<p>Oh ok i see. It's all new to me but it seems like the gulp command to copy the *.js would result in ALOT of js files being copied if grabbing the latest angular2 NPM no? Is that a concern at all? My project.json looks like:</p><p>"@angular/common": "2.0.0",<br>    "@angular/compiler": "2.0.0",<br>    "@angular/core": "2.0.0",<br>    "@angular/forms": "2.0.0",<br>    "@angular/http": "2.0.0",<br>    "@angular/platform-browser": "2.0.0",<br>    "@angular/platform-browser-dynamic": "2.0.0",<br>    "@angular/router": "3.0.0",<br>    "@angular/upgrade": "2.0.0",</p><p>And inside node_modules/@angular are dev, full and minified versions of all kinds. Thanks for all your help. Much appreciated!</p>

**Comment by JuergenGutsch** [comment]
Comment written by **gviuytgcv** on **21.09.2016 18:54:06**

<p>It is possible.. trying to figure it out myself. The benefit of a mixed mode is authentication, static views/pages. There were some nuget sample angular 1 MVC mixed mode samples. Would love to see an Angular2 one.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **agus** on **25.09.2016 20:55:31**

<p>Hello Jurgen, I got this error when trying create package.json.</p><p><a href="https://uploads.disquscdn.com/images/96d5e109d84c4b51d636ba6c19f08da6844bc022f93aac40c6642963001f415c.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/96d5e109d84c4b51d636ba6c19f08da6844bc022f93aac40c6642963001f415c.png">https://uploads.disquscdn.c...</a>  <a href="https://uploads.disquscdn.com/images/72b0422ad8de2451d52c0efa289ffdeb8246f66a317dacd22ee2b200f854327b.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/72b0422ad8de2451d52c0efa289ffdeb8246f66a317dacd22ee2b200f854327b.png">https://uploads.disquscdn.c...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **26.09.2016 08:23:25**

<p>Hi,<br>this looks like an network  issue. Either you were not connected to the internet or <a href="http://npm.org" rel="nofollow noopener" title="npm.org">npm.org</a> could not be reached...<br>Did you try to run npm install in the console? Open a console cd to your project, where the package.json is stored and type npm install</p>

**Comment by JuergenGutsch** [comment]
Comment written by **agus** on **26.09.2016 09:47:45**

<p>Thanks Jurgen, this problem because i'm not set the npm proxy</p>

**Comment by JuergenGutsch** [comment]
Comment written by **agus** on **26.09.2016 15:27:30**

<p>Jurgen&lt; i can't run the project.</p><p><a href="https://uploads.disquscdn.com/images/117fb3eaf0960f8b4cb3a7b4e9db281aa8fc2d446b772f481751fc3426292fe6.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/117fb3eaf0960f8b4cb3a7b4e9db281aa8fc2d446b772f481751fc3426292fe6.png">https://uploads.disquscdn.c...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **27.09.2016 11:11:58**

<p>Yea. VS has some issues finding all the NPM dependencies. Maybe because the VS tooling isn't finished yet. This is why I use the console to resolve the NPM dependencies. <br>npm install in the project folder should solve that build problem. But this error sign will still be there in VS.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **27.09.2016 11:17:54**

<p>Yes it is possible, but I don't really see any benefits. IMO: using token based auth is much more easier than trying to mix that stuff: <a href="http://asp.net-hacker.rocks/2016/09/22/web-api-authentication-in-aspnetcore-and-angular2.html" rel="nofollow noopener" title="http://asp.net-hacker.rocks/2016/09/22/web-api-authentication-in-aspnetcore-and-angular2.html">http://asp.net-hacker.rocks...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Maseeha** on **29.09.2016 08:26:19**

<p>After adding the given code below I am getting an error.. What do I do ?</p><p>app.Use(async (context, next) =&gt;<br>{<br>    await next();</p><p>    if (context.Response.StatusCode == 404<br>        &amp;&amp; !Path.HasExtension(context.Request.Path.Value))<br>    {<br>        context.Request.Path = "/index.html";<br>        await next();<br>    }<br>});</p><p><a href="https://uploads.disquscdn.com/images/b4f7f6dd534ce045052279df9702e9230a7ac15a283bc693b934d86f61b74815.png" rel="nofollow noopener" title="https://uploads.disquscdn.com/images/b4f7f6dd534ce045052279df9702e9230a7ac15a283bc693b934d86f61b74815.png">https://uploads.disquscdn.c...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **29.09.2016 10:16:07**

<p>Add "using System.IO;" at the top of the Startup.cs</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jono** on **29.09.2016 10:28:17**

<p>Ok, so this comment got waaaaaaay out of hand.</p><p>If I had a blog I may look at writing this up. Will look to github it soon.</p><p>Also some more things:</p><p>1. The angular packages can all be ~2.0.1 now<br>2. Do an 'npm install npm -g' to ensure npm is the latest version, otherwise it doesn't seem to like the angular dependencies<br>3. If you are behind a corporate firewall, create a .typingsrc file in c:\users\youname with the following:</p><p>{<br>  "proxy": "http://address:port",<br>  "rejectUnauthorized": false,<br>  "registryURL": "<a href="https://api.typings.org/" rel="nofollow noopener" title="https://api.typings.org/">https://api.typings.org/</a>"<br>}</p><p>4. If not obvious, main.ts should be in the wwwroot\app folder. The post says to for app.component.ts, but not main<br>5. The new version of angular does bootstrapping a bit differently:</p><p>main.ts should be:<br>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';<br>import { AppComponent } from './app.component';</p><p>const platform = platformBrowserDynamic();<br>platform.bootstrapModule(AppComponent);</p><p>Actually, we should be bootstrapping a module, and not a component:<br>import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';<br>import { AppModule } from './app.module';</p><p>const platform = platformBrowserDynamic();<br>platform.bootstrapModule(AppModule);</p><p>with app.module.ts:</p><p>import { NgModule } from '@angular/core';<br>import { BrowserModule } from '@angular/platform-browser';<br>import { AppComponent } from './app.component';</p><p>@NgModule({<br>    imports: [BrowserModule],<br>    declarations: [AppComponent],<br>    bootstrap: [AppComponent]<br>})</p><p>export class AppModule { }</p><p>6. I needed to manually remove some leftover packages that were not cleaned up from the package.json in this post, even though they were no longer referenced in the package.json</p><p>7. Your gulpfile (and index.html) uses libs, but systemjs.config uses lib<br>8. Your gulpfile copies angular folders to the libs route, so the systemjs.config is looking for them in the wrong place (libs/@angular). You can add a 'base' option to src to read into the node_modules folder from src.</p><p>gulp.task('restore', function () {<br>    gulp.src([<br>        './node_modules/@angular/**/*.js',<br>        './node_modules/angular-in-memory-web-api/*.js',<br>        './node_modules/rxjs/**/*.js',<br>        './node_modules/systemjs/dist/*.js',<br>        './node_modules/zone.js/dist/*.js',<br>        './node_modules/core-js/client/*.js',<br>        './node_modules/reflect-metadata/reflect.js',<br>        './node_modules/jquery/dist/*.js',<br>        './node_modules/bootstrap/dist/**/*.*'<br>    ], { base: './node_modules' }).pipe(gulp.dest('./wwwroot/libs/'));<br>});</p><p>9. Had lots of issues with getting a (SystemJS) Invalid Token &lt; sort of message. This was due to the MVC handler passing 404s back to index.html. Once I removed that I also found Traceur errors (could not load traceur, with a 404). No one has any information regarding it online, but it seems that using the angular umds is better than index.js:</p><p>packageNames.forEach(function (pkgName) {<br>        var p = pkgName.replace('@angular/', '');<br>        packages[pkgName] = { main: 'bundles/' + p + '.umd.js', defaultExtension: 'js' };<br>        //packages[pkgName] = { main: 'index.js', defaultExtension: 'js' };<br>    });</p>

**Comment by JuergenGutsch** [comment]
Comment written by **JT Nichols** on **29.09.2016 23:09:22**

<p>I'm glad I'm not the only one. I've tried the Angular 2 VS quickstart, a pluralsight project, and two other basic starter sites with VS and Angular2, all different errors around either VS, node or Angular2.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **TruthSayer** on **30.09.2016 00:07:38**

<p>Forget VS, it's not ready for the new stuff. VS code will give you hope.<br><a href="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html" rel="nofollow noopener" title="http://asp.net-hacker.rocks/2016/09/19/aspnetcore-and-angular2-using-dotnetcli-and-vscode.html">http://asp.net-hacker.rocks...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **30.09.2016 07:13:13**

<p>Most time the errors are because of the wrong versions in the package.json or a wrong typings.config. This post is done using Angular2 RC4. If you try to do it using the final version, it will fail, because the final version works different. <br>MSFT is not the only company doing breaking changes while RC ;)</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **30.09.2016 07:13:35**

<p>Yes, things are much more easier using VS Code</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jono** on **30.09.2016 13:40:27**

<p>I have made an Angular 2.0.1 template project now. Currently working through building a small app, to show folder-&gt;feature layout.</p><p><a href="https://github.com/nzcoward/NetCore-Angular" rel="nofollow noopener" title="https://github.com/nzcoward/NetCore-Angular">https://github.com/nzcoward...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Cristiano Dias** on **02.10.2016 22:30:31**

<p>Please, I little help would be of great value. I'm getting these errors when I try to run the app in browser:</p><p>GET http://localhost:52268/lib/@angular/platform-browser-dynamic/index.js 404 (Not Found) Zone.js<br>GET Error: Error: XHR error (404 Not Found) loading http://localhost:52268/lib/@angular/platform-browser-dynamic/index.js(…)<br>GET http://localhost:52268/lib/@angular/core/index.js 404 (Not Found) Zone.js</p><p>I just don't know what to do in this case.<br>Thanks in advance</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jono** on **03.10.2016 08:09:02**

<p>It could be that you are pointing at the 'lib' folder, but the gulpfile moves it to libs. If you check out my (shorter) comment, you will find a 2.0.1 template which is based off this post and should work for you. The longer comment details some things I came across in this post (some of which are 2.0.1 related, but some not).</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **03.10.2016 12:37:13**

<p>Hi Christiano, Jono is right. Please check the paths.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Lerry Mashaba** on **05.10.2016 11:28:39**

<p>Are  you planning to update this tutorial to Angular 2 final with .net core 1.0?</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **06.10.2016 07:59:35**

<p>Hi Larry, you are right. I should do it. Will do it this evening :)<br>It is already .NET Core 1.0, but I should update to 1.0.1</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Vamplife Eightyfour** on **10.10.2016 03:00:22**

<p>I just wanted to get something up online quick, I thought it would be nice to have MVC 6! &amp; ANGULAR. 2!!.. it's been almost a week..</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **12.10.2016 07:23:35**

<p>Hi Larry, <br>I just updated this tutorial and added a link to a working version.<br>Little later than expected, but I had some troubles with ASP.NET Core 1.0.1 in VS 2015</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Ryan Lee** on **13.10.2016 10:10:44**

<p>This might be off the topic but would you be able to share how to set up angular 2 working in one of .cshtml file in mvc5? Not .Net Core?</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **13.10.2016 15:27:35**

<p>It isn't really off-topic. Many thanks for that Question, Ryan :)</p><p>While using TypeScript, NPM and Gulp there are lot of files created or copied dynamically. <br>The classic Visual Studio project doesn't really support dynamically added files. Especially deploying using VS could be a problem, because VS only deploys files which are added to the project file. This means you need to add every single dynamically created file to the project, which is really a pain.<br>This is why I always use to go with a ASP.NET Core project.</p><p>Using Webpack instead of Gulp could be an option, but this is possible little more effort. This way reduces the amount of the needed dynamic fies to three to four files...</p><p>I will try it out and write a post about it, if it works.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Ryan Lee** on **15.10.2016 08:33:48**

<p>Thank you very much for your kind reply. I got most part working and it works when I put all the scripts in a HTML file but it fails to read the automatic js extension confit when I put the config file. I think I will try to follow the relevant steps you have mentioned here and try again. I will let you know if I make some progress.</p><p>Regards<br>Ryan</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **15.10.2016 08:51:47**

<p>This sounds like an problem with the paths in the systemjs.config.js. Maybe the path to the libs folder is wrong or the gulpfile wasn't executed...</p><p>Have you tried to have a look into that sample: <a href="https://github.com/JuergenGutsch/angular2-aspnetcore-vs" rel="nofollow noopener" title="https://github.com/JuergenGutsch/angular2-aspnetcore-vs">https://github.com/JuergenG...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **15.10.2016 08:59:29**

<p>I just updated that post to the latest versio ov Angular2. The most problem with this post was, that NPM fetches newer versions of the dependencies, because the package json has that sucking ^ bevore the version number.</p><p>So I updated this tutorial to the final angular2 version, removed the ^ from the package.json and uploaded a running sample to github: <a href="https://github.com/JuergenGutsch/angular2-aspnetcore-vs" rel="nofollow noopener" title="https://github.com/JuergenGutsch/angular2-aspnetcore-vs">https://github.com/JuergenG...</a></p>

**Comment by JuergenGutsch** [comment]
Comment written by **Ryan Lee** on **15.10.2016 09:00:41**

<p>That's part of what I am going to try out tonight :) I have 2 questions though. <br>1. Is gulp purely for generating js files?<br>2. The package.json file has dependency entries like angular2:2.0.0-beta-17 where as the example you have provided has @angular ( has @ ) . Would you be able to explain the difference ? <br>Regards<br>Ryan </p>

**Comment by JuergenGutsch** [comment]
Comment written by **Jürgen Gutsch** on **15.10.2016 10:43:45**

<p>1. Gulp is used to copy the js and css dependencies from /node_modules/ to /wwwroot/libs/, because node_modules should not be deployed<br>2. the package.json you mentioned is outdated, the new packages uses an @   in the name. See here: <a href="https://angular.io/docs/ts/latest/quickstart.html#!#create-and-configure" rel="nofollow noopener" title="https://angular.io/docs/ts/latest/quickstart.html#!#create-and-configure">https://angular.io/docs/ts/...</a> It is just about renamed NPM packages</p>

**Comment by JuergenGutsch** [comment]
Comment written by **A.T. Silva** on **15.10.2016 12:40:00**

<p>You mentioned adding the "&amp;&amp; gulp restore" but it is not shown in the code listing for the file "package.json". Is not on the GitHub either. Very helpful post though, saved me a lot of time, thank you.</p>

**Comment by JuergenGutsch** [comment]
Comment written by **Ryan Lee** on **19.10.2016 03:06:43**

<p>Thanks for the reply once again. I attempted to make this work again. Configured everything exactly how you set it up. It works on a normal html file, but it will fails when I put the relevant codes in an cshtml file. MVC routing prevents access to 'app', so I gave it a full path using "System.import('/app/main.js').catch(function (err) { console.error(err); });" but it fails to load the other ones. I understand what the problem is, mvc not allowing the request static files without extensions, however I don't really know how I can resolve this.</p>
